File: 1014-k-closest-points-to-origin/k-closest-points-to-origin.java
class Solution {
    public int[][] kClosest(int[][] points, int k) {
        k = points.length - k;
        quickSelect(0, points.length - 1, points, k);
        return Arrays.copyOfRange(points, k, points.length);
    }

    private void quickSelect(int left, int right, int[][] points, int k) {
        int pointer = left;
        int pivot = distance(points[right]);
        for (int i = left; i < right; i++) {
            int dis = distance(points[i]);
            if (distance(points[i]) >= pivot) {
                swap(i, pointer, points);
                pointer++;
            }
        }
        swap(pointer, right, points);
        if (pointer > k) {
            quickSelect(left, pointer - 1, points, k);
        } else if (pointer < k) {
            quickSelect(pointer + 1, right, points, k);
        }
        //return true;
    }

    private void swap(int i, int j, int[][] points) {
        int temp0 = points[i][0];
        int temp1 = points[i][1];
        points[i][0] = points[j][0];
        points[i][1] = points[j][1];
        points[j][0] = temp0;
        points[j][1] = temp1;
    }

    private int distance(int[] point) {
        return point[0] * point[0] + point[1] * point[1];
    }
}

File: 1014-k-closest-points-to-origin/k-closest-points-to-origin.py
class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        distance = [a*a + b*b for a, b in points]
        def quickselect(k, left, right):
            if left < right:
                pivot = partition(left, right)
                if pivot == k:
                    return points[:k]
                if pivot < k:
                    quickselect(k, pivot + 1, right)
                else:
                    quickselect(k, left, pivot - 1)

        def partition(left, right):
            pivot = distance[right]
            p1 = left
            for p2 in range(left, right):
                if distance[p2] <= pivot:
                    distance[p2], distance[p1] = distance[p1], distance[p2]
                    points[p2], points[p1] = points[p1], points[p2]
                    p1 += 1
            distance[right], distance[p1] = distance[p1], distance[right]
            points[right], points[p1] = points[p1], points[right]
            return p1
        quickselect(k, 0, len(points) - 1)
        return points[:k]